### ISSUE TITLE
Opentelemetry-jaeger: Problems using Collector Client in conjunection with Actix-Web

### ISSUE THREAD

**Issue by glademiller** [open-telemetry/opentelemetry-rust#106]
This is an amazing project and I am attempting to use it in conjunction with the tracing crate. I have posted my initialization code below. I am getting traces to my collector for things that are running in tokio outside of Actix but the actix-web handlers don't work. My best guess after digging through the code is the problem is due to the use of the blocking reqwest client.  For reference this is how I am instrumenting my actix-web handlers

```rust
#[tracing::instrument(skip(app))]
#[get("/health")]
pub async fn health(app: Data<Application>) -> Result<HttpResponse, ApplicationError> {
```
```
thread 'tokio-runtime-worker' panicked at 'Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks.', /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/enter.rs:38:5
```

```rust
use opentelemetry::{global, sdk, api::Provider};
use tracing_subscriber::layer::SubscriberExt;
use tracing_opentelemetry::OpenTelemetryLayer;

pub fn init_tracer(collector_endpoint: &str, collector_username: &str, collector_password: &str) -> std::io::Result<()> {
    let exporter: opentelemetry_jaeger::Exporter = opentelemetry_jaeger::Exporter::builder()
        .with_collector_endpoint(collector_endpoint)
        .with_collector_username(collector_username)
        .with_collector_password(collector_password)
        .with_process(opentelemetry_jaeger::Process {
            service_name: "service".to_string(),
            tags: vec![opentelemetry::api::core::KeyValue::new("environment", "dev")],
        })
        .init()
        .map_err(|e| std::io::Error::new(std::io::ErrorKind::Other, e))?;

    let provider = sdk::Provider::builder()
        .with_simple_exporter(exporter)
        .with_config(sdk::Config {
            default_sampler: Box::new(sdk::Sampler::Always),
            ..Default::default()
        })
        .build();


    global::set_provider(provider);
    let tracer = global::trace_provider().get_tracer("service");
    let telemetry = OpenTelemetryLayer::with_tracer(tracer);
    let subscriber = tracing_subscriber::Registry::default().with(telemetry);

    tracing::subscriber::set_global_default(subscriber).expect("Could not set global subscriber");

    tracing::trace_span!("tracing_startup");

    Ok(())
}
```

**Comment by jtescher** [comment]
Hm this code should not be trying to start another tokio runtime. Can you include the output with  `RUST_BACKTRACE=1`?

**Comment by glademiller** [comment]
Reqwest's blocking client does althought I'm not entirely sure how the conflict is happening.

https://github.com/seanmonstar/reqwest/blob/master/src/blocking/client.rs#L696-L732

I will get a full stack trace.

**Comment by glademiller** [comment]
```
thread 'tokio-runtime-worker' panicked at 'Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks.', /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/enter.rs:38:5
stack backtrace:
   0:     0x5615fc153c94 - backtrace::backtrace::libunwind::trace::h5ae3454ebe40b621
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.44/src/backtrace/libunwind.rs:86
   1:     0x5615fc153c94 - backtrace::backtrace::trace_unsynchronized::hbf6f267a483fb2c0
                               at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.44/src/backtrace/mod.rs:66
   2:     0x5615fc153c94 - std::sys_common::backtrace::_print_fmt::h37755015d94b9085
                               at src/libstd/sys_common/backtrace.rs:78
   3:     0x5615fc153c94 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h0561c8cd3569b40d
                               at src/libstd/sys_common/backtrace.rs:59
   4:     0x5615fc1774cc - core::fmt::write::h86a597a35d48a212
                               at src/libcore/fmt/mod.rs:1063
   5:     0x5615fc14c403 - std::io::Write::write_fmt::h73d9dc0f29eefe48
                               at src/libstd/io/mod.rs:1426
   6:     0x5615fc1564f5 - std::sys_common::backtrace::_print::h6c60c5ee09af0233
                               at src/libstd/sys_common/backtrace.rs:62
   7:     0x5615fc1564f5 - std::sys_common::backtrace::print::hf9d1db01634f9b2a
                               at src/libstd/sys_common/backtrace.rs:49
   8:     0x5615fc1564f5 - std::panicking::default_hook::{{closure}}::h1db9298dd213ae95
                               at src/libstd/panicking.rs:204
   9:     0x5615fc156242 - std::panicking::default_hook::h3272faf6cc91fd2e
                               at src/libstd/panicking.rs:224
  10:     0x5615fc156b52 - std::panicking::rust_panic_with_hook::hb976084785e50594
                               at src/libstd/panicking.rs:470
  11:     0x5615fc131761 - std::panicking::begin_panic::hbda02cefef63ce4a
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/panicking.rs:397
  12:     0x5615fc0adb70 - tokio::runtime::enter::enter::h95635d966147e630
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/enter.rs:38
  13:     0x5615fc048bfc - tokio::runtime::blocking::shutdown::Receiver::wait::h2e02ada088ce9600
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/blocking/shutdown.rs:44
  14:     0x5615fc04e6fd - tokio::runtime::blocking::pool::BlockingPool::shutdown::hd052913c0aa77c94
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/blocking/pool.rs:132
  15:     0x5615fc04e7ec - <tokio::runtime::blocking::pool::BlockingPool as core::ops::drop::Drop>::drop::hed0c06b4a6012ceb
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/blocking/pool.rs:138
  16:     0x5615fc0a00b7 - core::ptr::drop_in_place::hca9693e13d3a81a8
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libcore/ptr/mod.rs:177
  17:     0x5615fc09dcda - core::ptr::drop_in_place::h6d13b909570362c6
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libcore/ptr/mod.rs:177
  18:     0x5615fb1cfc5f - reqwest::blocking::wait::enter::h755bd5c440e7f2ff
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/reqwest-0.10.4/src/blocking/wait.rs:74
  19:     0x5615fb1ce9f3 - reqwest::blocking::wait::timeout::hd02fb2200f02ebb3
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/reqwest-0.10.4/src/blocking/wait.rs:14
  20:     0x5615fb1e80a2 - reqwest::blocking::client::ClientHandle::execute_request::hd3d1c1b826c99c21
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/reqwest-0.10.4/src/blocking/client.rs:767
  21:     0x5615fb1e4e02 - reqwest::blocking::client::Client::execute::h345b84ea43d6435f
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/reqwest-0.10.4/src/blocking/client.rs:623
  22:     0x5615fb32237d - reqwest::blocking::request::RequestBuilder::send::ha0c94d2de1cf559f
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/reqwest-0.10.4/src/blocking/request.rs:522
  23:     0x5615fb1857e8 - <opentelemetry_jaeger::transport::http::THttpChannel as std::io::Write>::flush::hdc2cb4bdb4545aad
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/opentelemetry-jaeger-0.3.0/src/transport/http.rs:127
  24:     0x5615fb19da07 - <thrift::transport::WriteHalf<C> as std::io::Write>::flush::h3f1c648e52538a71
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/thrift-0.13.0/src/transport/mod.rs:175
  25:     0x5615fb19f393 - <thrift::protocol::binary::TBinaryOutputProtocol<T> as thrift::protocol::TOutputProtocol>::flush::h9101c767b0c3b3f6
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/thrift-0.13.0/src/protocol/binary.rs:433
  26:     0x5615fb184690 - <opentelemetry_jaeger::collector::CollectorSyncClientHttp as opentelemetry_jaeger::thrift::jaeger::TCollectorSyncClient>::submit_batches::{{closure}}::{{closure}}::h1807b5b401a39737
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/opentelemetry-jaeger-0.3.0/src/collector.rs:64
  27:     0x5615fb18830b - core::result::Result<T,E>::and_then::h512fcfe71d49eeb7
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libcore/result.rs:727
  28:     0x5615fb184742 - <opentelemetry_jaeger::collector::CollectorSyncClientHttp as opentelemetry_jaeger::thrift::jaeger::TCollectorSyncClient>::submit_batches::{{closure}}::h8c6319569d25728f
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/opentelemetry-jaeger-0.3.0/src/collector.rs:62
  29:     0x5615fb182f6b - core::iter::adapters::map_fold::{{closure}}::h307aa34d199a78a3
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libcore/iter/adapters/mod.rs:772
  30:     0x5615fb1824b9 - core::iter::traits::iterator::Iterator::fold::ok::{{closure}}::h87b2b65ebf9c12f6
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libcore/iter/traits/iterator.rs:2001
  31:     0x5615fb179faf - core::iter::traits::iterator::Iterator::try_fold::h5caeff8efb200a4d
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libcore/iter/traits/iterator.rs:1877
  32:     0x5615fb179d10 - core::iter::traits::iterator::Iterator::fold::h541e4ce1c5455d8b
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libcore/iter/traits/iterator.rs:2004
  33:     0x5615fb1837d0 - <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::fold::h5f9560dc521d5af5
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libcore/iter/adapters/mod.rs:812
  34:     0x5615fb182793 - core::iter::traits::iterator::Iterator::for_each::h84fa7e8fbb204f24
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libcore/iter/traits/iterator.rs:658
  35:     0x5615fb17c9aa - <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T,I>>::spec_extend::hc190c5b8ced637a2
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/liballoc/vec.rs:2061
  36:     0x5615fb17d5e5 - <alloc::vec::Vec<T> as alloc::vec::SpecExtend<T,I>>::from_iter::hfcf807a6e2d459a5
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/liballoc/vec.rs:2041
  37:     0x5615fb17e509 - <alloc::vec::Vec<T> as core::iter::traits::collect::FromIterator<T>>::from_iter::hf17f59485a4c173f
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/liballoc/vec.rs:1926
  38:     0x5615fb18267b - core::iter::traits::iterator::Iterator::collect::hd06a7dc7dfef4df4
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libcore/iter/traits/iterator.rs:1659
  39:     0x5615fb184630 - <opentelemetry_jaeger::collector::CollectorSyncClientHttp as opentelemetry_jaeger::thrift::jaeger::TCollectorSyncClient>::submit_batches::h5199bae91d45f21a
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/opentelemetry-jaeger-0.3.0/src/collector.rs:58
  40:     0x5615fb1813fa - opentelemetry_jaeger::uploader::BatchUploader::upload::h37025321619da937
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/opentelemetry-jaeger-0.3.0/src/uploader.rs:27
  41:     0x5615fb189815 - <opentelemetry_jaeger::Exporter as opentelemetry::exporter::trace::SpanExporter>::export::h84b04bb7cc922256
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/opentelemetry-jaeger-0.3.0/src/lib.rs:162
  42:     0x5615fb517ffd - <opentelemetry::sdk::trace::span_processor::SimpleSpanProcessor as opentelemetry::api::trace::span_processor::SpanProcessor>::on_end::hf00175e875b0cf35
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/opentelemetry-0.4.0/src/sdk/trace/span_processor.rs:124
  43:     0x5615fb5092ca - <opentelemetry::sdk::trace::span::SpanInner as core::ops::drop::Drop>::drop::h3f3ebb15211af09c
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/opentelemetry-0.4.0/src/sdk/trace/span.rs:176
  44:     0x5615fb510c67 - core::ptr::drop_in_place::h99563a1efb69699c
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libcore/ptr/mod.rs:177
  45:     0x5615fb51318d - alloc::sync::Arc<T>::drop_slow::haf8032cd49039f03
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/liballoc/sync.rs:739
  46:     0x5615fb513886 - <alloc::sync::Arc<T> as core::ops::drop::Drop>::drop::h767c79e0e6babe94
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/liballoc/sync.rs:1249
  47:     0x5615fb51089b - core::ptr::drop_in_place::h7464c0b0656153c2
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libcore/ptr/mod.rs:177
  48:     0x5615fb50fcf2 - core::ptr::drop_in_place::h13207b0129e3cbdc
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libcore/ptr/mod.rs:177
  49:     0x5615fb51026d - core::ptr::drop_in_place::h2c4a319853c60d32
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libcore/ptr/mod.rs:177
  50:     0x5615fb5110fb - core::ptr::drop_in_place::hbac15fe93c378615
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libcore/ptr/mod.rs:177
  51:     0x5615faded4f0 - <tracing_opentelemetry::layer::OpenTelemetryLayer<S,T> as tracing_subscriber::layer::Layer<S>>::on_close::h9423969099141dc1
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-opentelemetry-0.3.1/src/layer.rs:259
  52:     0x5615fb08328a - <tracing_subscriber::layer::Layered<L,S> as tracing_core::subscriber::Subscriber>::try_close::h4369b8aeee71c770
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-subscriber-0.2.5/src/layer.rs:642
  53:     0x5615fb4aab65 - tracing_core::dispatcher::Dispatch::try_close::hce56702fdee6e228
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-core-0.1.10/src/dispatcher.rs:559
  54:     0x5615fb4aa541 - <tracing::span::Span as core::ops::drop::Drop>::drop::hbc3c837aac44e1c4
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tracing-0.1.13/src/span.rs:910
  55:     0x5615fb4aa257 - core::ptr::drop_in_place::h51b33dff2468443c
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libcore/ptr/mod.rs:177
  56:     0x5615fb0d3870 - core::ptr::drop_in_place::h22693b0b2d22a149
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libcore/ptr/mod.rs:177
  57:     0x5615fb00d399 - crm_auth::db::gateway::CosmosGateway::refresh::{{closure}}::hdf7c296a08824cad
                               at src/db/gateway.rs:54
  58:     0x5615fb07911e - <std::future::GenFuture<T> as core::future::future::Future>::poll::h657a043e699b21f3
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/future.rs:44
  59:     0x5615fb0724f8 - std::future::poll_with_tls_context::hb8b0fa9c61e78312
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/future.rs:102
  60:     0x5615fb00b557 - crm_auth::db::gateway::CosmosGateway::start::{{closure}}::h7e2765183efecf39
                               at src/db/gateway.rs:44
  61:     0x5615fb07a95a - <std::future::GenFuture<T> as core::future::future::Future>::poll::h92a7473dda409b99
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/future.rs:44
  62:     0x5615faf20cd9 - tokio::runtime::task::core::Core<T,S>::poll::{{closure}}::h75752c30f8bc9555
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/task/core.rs:173
  63:     0x5615fb092167 - tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut::h53bb5b575726dea9
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/loom/std/unsafe_cell.rs:14
  64:     0x5615faf1fbf3 - tokio::runtime::task::core::Core<T,S>::poll::hb84acbfd22fc9146
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/task/core.rs:158
  65:     0x5615fae08e78 - tokio::runtime::task::harness::Harness<T,S>::poll::{{closure}}::h640efaf7d4f0cfd8
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/task/harness.rs:113
  66:     0x5615fb0cb080 - core::ops::function::FnOnce::call_once::h645798ed059e9bb8
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libcore/ops/function.rs:232
  67:     0x5615faedd9ab - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h079538b07c126556
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/panic.rs:318
  68:     0x5615fae5a99a - std::panicking::try::do_call::h6b6414819ca8037b
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/panicking.rs:303
  69:     0x5615fc15ba97 - __rust_maybe_catch_panic
                               at src/libpanic_unwind/lib.rs:86
  70:     0x5615fae597c3 - std::panicking::try::hf880899786316092
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/panicking.rs:281
  71:     0x5615faede66a - std::panic::catch_unwind::h7ff7da4dc56d8e7a
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/panic.rs:394
  72:     0x5615fae075e1 - tokio::runtime::task::harness::Harness<T,S>::poll::hf7a5e479f2ad1242
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/task/harness.rs:89
  73:     0x5615fae1fb80 - tokio::runtime::task::raw::poll::hcbb02f36c7e5eea9
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/task/raw.rs:104
  74:     0x5615fc067dbf - tokio::runtime::task::raw::RawTask::poll::h880adbf08bb6942a
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/task/raw.rs:66
  75:     0x5615fc0a4141 - tokio::runtime::task::Notified<S>::run::h67d1a71617b294e4
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/task/mod.rs:169
  76:     0x5615fc03b7a2 - tokio::runtime::thread_pool::worker::Context::run_task::{{closure}}::hfcdee52bac064da9
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/thread_pool/worker.rs:353
  77:     0x5615fc0a3cb4 - tokio::coop::budget::{{closure}}::ha4f19aed9a62c17b
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/coop.rs:97
  78:     0x5615fc084745 - std::thread::local::LocalKey<T>::try_with::hac78a4af3c5b8b5f
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/thread/local.rs:262
  79:     0x5615fc082459 - std::thread::local::LocalKey<T>::with::h455cf9317250ba49
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/thread/local.rs:239
  80:     0x5615fc03b679 - tokio::coop::budget::he28144f2e15119ac
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/coop.rs:79
  81:     0x5615fc03b679 - tokio::runtime::thread_pool::worker::Context::run_task::h698098b2c7ba974f
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/thread_pool/worker.rs:352
  82:     0x5615fc03b23d - tokio::runtime::thread_pool::worker::Context::run::h3d9836f3ee3591c0
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/thread_pool/worker.rs:331
  83:     0x5615fc03ae93 - tokio::runtime::thread_pool::worker::run::{{closure}}::h11b99bd90a701f91
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/thread_pool/worker.rs:309
  84:     0x5615fc066d26 - tokio::macros::scoped_tls::ScopedKey<T>::set::h84f815de8e2431fe
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/macros/scoped_tls.rs:64
  85:     0x5615fc03ad91 - tokio::runtime::thread_pool::worker::run::h4c9289154c675bc8
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/thread_pool/worker.rs:306
  86:     0x5615fc03abeb - tokio::runtime::thread_pool::worker::Launch::launch::{{closure}}::hbcd435d3c846b680
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/thread_pool/worker.rs:285
  87:     0x5615fc0680a6 - <tokio::runtime::blocking::task::BlockingTask<T> as core::future::future::Future>::poll::h075c6fa328ab2ef6
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/blocking/task.rs:38
  88:     0x5615fc0bd10e - tokio::runtime::task::core::Core<T,S>::poll::{{closure}}::h1bb8a872ddb8d959
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/task/core.rs:173
  89:     0x5615fc06f12b - tokio::loom::std::unsafe_cell::UnsafeCell<T>::with_mut::hf8f34de3dd231c57
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/loom/std/unsafe_cell.rs:14
  90:     0x5615fc0bcf4e - tokio::runtime::task::core::Core<T,S>::poll::h7f47d9776b61a3ed
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/task/core.rs:158
  91:     0x5615fc04cf92 - tokio::runtime::task::harness::Harness<T,S>::poll::{{closure}}::h89c36a2d3e614ebb
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/task/harness.rs:113
  92:     0x5615fc09b140 - core::ops::function::FnOnce::call_once::hcbf2d2332eee5a98
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libcore/ops/function.rs:232
  93:     0x5615fc0c1a2c - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::h784a9082b3e2695f
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/panic.rs:318
  94:     0x5615fc069a5c - std::panicking::try::do_call::h8180d9ff6f8272c9
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/panicking.rs:303
  95:     0x5615fc15ba97 - __rust_maybe_catch_panic
                               at src/libpanic_unwind/lib.rs:86
  96:     0x5615fc0697f3 - std::panicking::try::h5e96790d2aeae40c
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/panicking.rs:281
  97:     0x5615fc0c1b4b - std::panic::catch_unwind::h6a1aff88a6344940
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/panic.rs:394
  98:     0x5615fc04cabe - tokio::runtime::task::harness::Harness<T,S>::poll::h32be936f0f270918
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/task/harness.rs:89
  99:     0x5615fc067f12 - tokio::runtime::task::raw::poll::h0cb7eded569a594b
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/task/raw.rs:104
 100:     0x5615fc067dbf - tokio::runtime::task::raw::RawTask::poll::h880adbf08bb6942a
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/task/raw.rs:66
 101:     0x5615fc0a40d1 - tokio::runtime::task::Notified<S>::run::h136f01bc09c96f9c
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/task/mod.rs:169
 102:     0x5615fc04f371 - tokio::runtime::blocking::pool::Inner::run::h1d2c249c1ca38eed
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/blocking/pool.rs:230
 103:     0x5615fc04f0be - tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}::{{closure}}::h2493a51f59c05272
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/blocking/pool.rs:210
 104:     0x5615fc0bc0b5 - tokio::runtime::context::enter::h978eaf032a4b1439
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/context.rs:72
 105:     0x5615fc095cef - tokio::runtime::handle::Handle::enter::h73561aaa67b494b8
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/handle.rs:71
 106:     0x5615fc04f14f - tokio::runtime::blocking::pool::Spawner::spawn_thread::{{closure}}::hb7209b24c9dd151b
                               at /home/glade/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-0.2.20/src/runtime/blocking/pool.rs:209
 107:     0x5615fc052932 - std::sys_common::backtrace::__rust_begin_short_backtrace::h4812d0c2e630c669
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/sys_common/backtrace.rs:130
 108:     0x5615fc04a7c3 - std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}::hfa486a9b9ac5ecd0
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/thread/mod.rs:475
 109:     0x5615fc0c1a91 - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hf439d270d7e31f25
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/panic.rs:318
 110:     0x5615fc069b71 - std::panicking::try::do_call::hc7ae10ea9d02e6fd
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/panicking.rs:303
 111:     0x5615fc15ba97 - __rust_maybe_catch_panic
                               at src/libpanic_unwind/lib.rs:86
 112:     0x5615fc0696a8 - std::panicking::try::h34513ef692dd9a26
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/panicking.rs:281
 113:     0x5615fc0c1b93 - std::panic::catch_unwind::heb93a87a0bda125b
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/panic.rs:394
 114:     0x5615fc04a5ab - std::thread::Builder::spawn_unchecked::{{closure}}::hfcecf89590c4261e
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libstd/thread/mod.rs:474
 115:     0x5615fc09ad9f - core::ops::function::FnOnce::call_once{{vtable.shim}}::h2968a7b25731ce53
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/libcore/ops/function.rs:232
 116:     0x5615fc145d2f - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::h553ef812d1929d1b
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/liballoc/boxed.rs:1017
 117:     0x5615fc15abbd - <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::h51b51bce029ae491
                               at /rustc/4fb7144ed159f94491249e86d5bbd033b5d60550/src/liballoc/boxed.rs:1017
 118:     0x5615fc15abbd - std::sys_common::thread::start_thread::hca943f45f04c8e46
                               at src/libstd/sys_common/thread.rs:13
 119:     0x5615fc15abbd - std::sys::unix::thread::Thread::new::thread_start::h352e8a5875b189ee
                               at src/libstd/sys/unix/thread.rs:80
 120:     0x7f5a3548d6db - start_thread
 121:     0x7f5a34f9e88f - __clone
 122:                0x0 - <unknown>
```

**Comment by glademiller** [comment]
Looking at the stack trace the error is being caused here.

https://github.com/seanmonstar/reqwest/blob/master/src/blocking/wait.rs#L66-L76

**Comment by glademiller** [comment]
Looking closer at the stack trace the error is not actually due to actix-web in this case. The start function referenced uses tokio:spawn to run something at an interval. The refresh function is instrumented and is made as part of this loop.

**Comment by jtescher** [comment]
Ah yes digging into this it looks like the blocking and non-blocking APIs of tokio and reqwest do not play nicely together. Basically if your app uses async / await and futures you must currently use the thrift api with the agent endpoint instead of the collector client.

**Comment by LukeMathWalker** [comment]
I am having exactly the same issue - any chance we could use the async interface od Reqwest for the collector, instead of the blocking one?

**Comment by jtescher** [comment]
@LukeMathWalker yeah that would be more ideal, have to think through the api for async exporters. Have an issue open currently for that specifically https://github.com/open-telemetry/opentelemetry-rust/issues/112

**Comment by jtescher** [comment]
Resolved in #121
